File: src/main/java/org/jcodec/containers/mp4/boxes/PixelAspectExt.java
package org.jcodec.containers.mp4.boxes;

import org.jcodec.common.model.Rational;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org )
 * This software is distributed under FreeBSD License
 * 
 * Pixel aspect ratio video sample entry extension
 * 
 * @author The JCodec project
 *
 */
public class PixelAspectExt extends Box {
    private int hSpacing;
    private int vSpacing;

    public PixelAspectExt(Header header) {
        super(header);
    }

    public static PixelAspectExt createPixelAspectExt(Rational par) {
        PixelAspectExt pasp = new PixelAspectExt(new Header(fourcc()));
        pasp.hSpacing = par.getNum();
        pasp.vSpacing = par.getDen();
        return pasp;
    }
    
    public void parse(ByteBuffer input) {
        hSpacing = input.getInt();
        vSpacing = input.getInt();
    }

    protected void doWrite(ByteBuffer out) {
        out.putInt(hSpacing);
        out.putInt(vSpacing);
    }
    
    @Override
    public int estimateSize() {
        return 16;
    }

    public int gethSpacing() {
        return hSpacing;
    }

    public int getvSpacing() {
        return vSpacing;
    }
    
    public Rational getRational() {
        return new Rational(hSpacing, vSpacing);
    }

    public static String fourcc() {
        return "pasp";
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/PixelAspectExt.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.common.model.Rational
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org )
 * This software is distributed under FreeBSD License
 *
 * Pixel aspect ratio video sample entry extension
 *
 * @author The JCodec project
 */
class PixelAspectExt(header: Header) : Box(header) {
    private var hSpacing = 0
    private var vSpacing = 0
    override fun parse(buf: ByteBuffer) {
        hSpacing = buf.int
        vSpacing = buf.int
    }

    override fun doWrite(out: ByteBuffer) {
        out.putInt(hSpacing)
        out.putInt(vSpacing)
    }

    override fun estimateSize(): Int {
        return 16
    }

    fun gethSpacing(): Int {
        return hSpacing
    }

    fun getvSpacing(): Int {
        return vSpacing
    }

    val rational: Rational
        get() = Rational(hSpacing, vSpacing)

    companion object {
        @JvmStatic
        fun createPixelAspectExt(par: Rational): PixelAspectExt {
            val pasp = PixelAspectExt(Header(fourcc()))
            pasp.hSpacing = par.getNum()
            pasp.vSpacing = par.getDen()
            return pasp
        }

        @JvmStatic
        fun fourcc(): String {
            return "pasp"
        }
    }
}